lib/Parse/ParseDecl.cpp

```diff
@@ -1530,7 +1529,133 @@ bool Parser::parseNewDeclAttribute(DeclAttributes &Attributes, SourceLoc AtLoc,
                                                  /*Implicit=*/false));
     break;
   }
-
+  case DAK_OriginallyDefinedIn: {
+    auto LeftLoc = Tok.getLoc();
+    if (!consumeIf(tok::l_paren)) {
+      diagnose(Loc, diag::attr_expected_lparen, AttrName,
+               DeclAttribute::isDeclModifier(DK));
+      return false;
+    }
+    SourceLoc RightLoc;
+    enum class NextSegmentKind: uint8_t {
+      ModuleName = 0,
+      VersionNumber,
+      End,
+    };
+    NextSegmentKind NK = NextSegmentKind::ModuleName;
+    StringRef OriginalModuleName;
+    PlatformKind Platform = PlatformKind::none;
+    Optional<llvm::VersionTuple> MovedVersion;
+    StringRef AttrName = "@_originalDefinedIn";
+    if (parseList(tok::r_paren, LeftLoc, RightLoc, false,
+                  diag::originally_defined_in_missing_rparen,
+                  SyntaxKind::Unknown, [&]() -> ParserStatus {
```

Comment:
`parseList()` doesn't provide much benefit here. Could you just parse them sequentially? Like 
```cpp
// Parse 'module' ':'.
if (!Tok.is(tok::identifier) || Tok.getText != "module" || peekToken().is(tok::colon)) {
  diagnose(Tok, diag::originally_defined_in_need_original_module_name);
  return false;
}
consumeToken(tok::identifier);
consumeToken(tok::colon);

// Parse the original module name; string_literal.
Optional<StringRef> NameOp;
if (Tok.is(tok::string_literal)) {
  if (auto NameOp = getStringLiteralIfNotInterpolated(Tok.getLoc(), "original module name"))
    OriginalModuleName = *NameOp;
}
if (!NameOp.hasValue()) {
  diagnose(Tok, diag::originally_defined_in_need_nonempty_module_name);
  return false;
}

// Parse ','.
if (!consumeIf(tok::comma))
  diagnose(Tok, diag::expected_separator, ",");
  return false;
}

// Parse the platform name and the version number; identifier 
if (!Tok.is(tok::identifier || !peekToken.isAny(tok::floating_literal, tok::integer_literal)) {
  diagnose(Tok, diag::attr_availability_expected_version, AttrName);
  return false;
}
...
```

Reply:
hmm, It seems we do need to use `parseList` here since we should allow multiple platform & version numbers like `@_orginallyDefined(module: "Foo", OSX 10.7, iOS 7.0)`. Will change the parser logic.

Reply:
Still `module: "<module name>" ,` part can be parsed outside `parseList`. So the remaining elements are only `platform version`s. You don't need to maintain `NK` state and the switch.